### DISCUSSION TITLE
insertOrUpdate with auto-inc primary key and unique key

### DISCUSSION THREAD

**Discussion by jcrist1** [slick/slick#2700] [Questions]
Hi There,
First, thanks for the great work on slick. I know OS dev can be a thankless job at sometimes and a lot of feedback is critical because people only reach out when something isn't working. In general we have been using slick effectively and appreciate how it makes SQL composable, and gives us compile time guarantees.

But we recently upgraded to slick 3.4.1 from 3.3.3, and noticed a change in behavior in the `insertOrUpdate` method with MariaDB.
We have a schema like:
```scala
  class EntityTable(tag: Tag) extends Table[Entity](tag, "entity") {
    def id         = column[EntityId]("id", O.PrimaryKey, O.AutoInc)
    def externalId     = column[ExternalId]("external_id", O.Unique)
    def updated = column[DateTime](createdColumn)
    
    def * =
      (id, externalId, created).<>(Entity.tupled, Entity.unapply)
  }
```

Prior to 3.4.1 we would use `insertOrUpdate` and the following would yield `Futures.successful(true)` as `res2` and `res1` should both be `EntityId(1)` (from empty).
In 3.4.1 `res2` will now be `EntityId(2)`

```scala
for {
  _ <- db.run(EntityTable.insertOrUpdate(Entity(EntityId(0), ExternalId("maybeAnEmail@address.com"), DateTime.now())))
  res1 <- db.run(EntityTable.filter(_.externalId === ExternalId("maybeAnEmail@address.com")).map(_.id).result())
  _ <- db.run(EntityTable.insertOrUpdate(Entity(EntityId(0), ExternalId("maybeAnEmail@address.com"), DateTime.now())))
  res2 <- db.run(EntityTable.filter(_.externalId === ExternalId("maybeAnEmail@address.com")).map(_.id).result())
} yield {
  res1 == res2
}
```

This seems to be due to the changes in [this PR](https://github.com/slick/slick/pull/2201) [this comment](https://github.com/slick/slick/pull/2201/files#r604655407) 
highlights the test that demonstrates that we now try to update the auto increment Id column. For MariaDB [it seems](https://mariadb.com/kb/en/insert-on-duplicate-key-update/) 
like trying to update the auto increment column will trigger it to be set to the next auto-increment value.

This is a bit unfortunate for us, because we usually use the internal id as a foreign key in a lot of places, and don't want to cascade update on each update to some related data,
as this will slow down inserts. What we would like to continue to do is 
* if no matching unique data insert with new auto-increment primary key
* if matching unique data update everything except the primary key
* if updating based on primary key and collide with unique data, throw an exception. Here we would use the `update` method anyway.

How could we do this in 3.4 and above?

**Comment by nafg** [comment]
First, I would say try the latest milestone release.

If that doesn't help @jkugiya might be able to help.

**Reply by jkugiya** [reply]
> - if no matching unique data insert with new auto-increment primary key
> - if matching unique data update everything except the primary key
> - if updating based on primary key and collide with unique data, throw an exception. Here we would use the update method anyway.

Thanks for the feedback. Let me check a few things to make sure I understand the content correctly.
Regarding the third, this probably wasn't the way it worked in previous versions. (It was not updated without throwing an exception).
So you are saying that an exception should be thrown?

**Reply by jkugiya** [reply]
This seems difficult to deal with because I don't think there is a way to use MySQL's upsert syntax to make an error when a particular column is duplicated. (There is a syntax that skips it rather than making it an error.)

It is difficult to determine the requirements, whether your example program should also be an error due to duplication or whether we should update the date.

Basically, the approach would be similar to what we are doing with user-side upsert.
You can extend MyProfile to do user-side upserts, but be aware that some tests will not pass in that case.

**Reply by jkugiya** [reply]
The fact is that MySQL has several upsert options, so a possible approach would be for the library to provide several options, but it would be complicated for the library. What do you think?

**Comment by jkugiya** [comment]
#2785

Do you think this fix will work?
